File: Trabajo Parcial/Parcial.cpp
#include <iostream>
#include <fstream>
#include <cstring>
#include <vector>

using namespace std;

// Estructura de los libros
struct Libro
{
    string titulo;
    string autor;
    int anioEdicion;
    string editorial;
    string ISBN;
    int numPaginas;
    Libro *siguiente;
};

// Estructura de los lectores
struct Lector
{
    string nombre;
    string dni;
    string libroSolicitado;
    Lector *siguiente;
};

Libro *cabeza = nullptr;
Lector *frente = nullptr;
Lector *final = nullptr;
string historial[100];
int indiceHistorial = 0;

// Funciones
void agregarHistorial(const string &accion)
{
    if (indiceHistorial < 100)
    {
        historial[indiceHistorial++] = accion;
    }
}

void cargarLibrosDesdeArchivo(const string &nombreArchivo)
{
    ifstream archivo(nombreArchivo);
    if (!archivo.is_open())
    {
        cerr << "Error: No se pudo abrir el archivo " << nombreArchivo << endl;
        return;
    }
    while (archivo)
    {
        string titulo, autor, editorial, ISBN;
        int anioEdicion, numPaginas;
        getline(archivo, titulo, ',');
        getline(archivo, autor, ',');
        archivo >> anioEdicion;
        archivo.ignore();
        getline(archivo, editorial, ',');
        getline(archivo, ISBN, ',');
        archivo >> numPaginas;
        archivo.ignore();

        if (!titulo.empty())
        {
            Libro *nuevoLibro = new Libro{titulo, autor, anioEdicion, editorial, ISBN, numPaginas, cabeza};
            cabeza = nuevoLibro;
        }
    }
    archivo.close();
}

void agregarLibro(const string &titulo, const string &autor, int anioEdicion, const string &editorial, const string &ISBN, int numPaginas)
{
    Libro *nuevoLibro = new Libro{titulo, autor, anioEdicion, editorial, ISBN, numPaginas, cabeza};
    cabeza = nuevoLibro;
    agregarHistorial("Agregar libro: " + titulo);
}

void mostrarLibros()
{
    if (!cabeza)
    {
        cout << "No hay libros en la biblioteca." << endl;
        return;
    }
    Libro *actual = cabeza;
    while (actual)
    {
        cout << "Titulo: " << actual->titulo << ", Autor: " << actual->autor
             << ", Anio: " << actual->anioEdicion << ", Editorial: " << actual->editorial
             << ", ISBN: " << actual->ISBN << ", Paginas: " << actual->numPaginas << endl;
        actual = actual->siguiente;
    }
}

void ordenarLibros()
{
    if (!cabeza || !cabeza->siguiente)
    {
        cout << "No hay suficientes libros para ordenar." << endl;
        return;
    }

    bool intercambiado;
    do
    {
        intercambiado = false;
        Libro *actual = cabeza;
        Libro *anterior = nullptr;
        Libro *siguiente = cabeza->siguiente;

        while (siguiente)
        {
            if (actual->titulo > siguiente->titulo)
            {
                
                actual->siguiente = siguiente->siguiente;
                siguiente->siguiente = actual;

                if (anterior)
                {
                    anterior->siguiente = siguiente;
                }
                else
                {
                    cabeza = siguiente;
                }

                intercambiado = true;
                anterior = siguiente;
                siguiente = actual->siguiente;
            }
            else
            {

                anterior = actual;
                actual = siguiente;
                siguiente = siguiente->siguiente;
            }
        }
    } while (intercambiado);

    cout << "Libros ordenados por titulo." << endl;
}

void eliminarLibroDeLista(const string &libroSolicitado)
{
    Libro *actual = cabeza;
    Libro *anterior = nullptr;

    while (actual)
    {
        if (actual->titulo == libroSolicitado)
        {
            if (anterior)
            {
                anterior->siguiente = actual->siguiente;
            }
            else
            {
                cabeza = actual->siguiente;
            }
            delete actual;
            return;
        }
        anterior = actual;
        actual = actual->siguiente;
    }
}

void eliminarLibroDelArchivo(const string &libroSolicitado)
{
    ifstream archivo("biblioteca.txt");
    ofstream temp("temp.txt");
    string titulo, autor, editorial, ISBN;
    int anioEdicion, numPaginas;

    while (getline(archivo, titulo, ','))
    {
        getline(archivo, autor, ',');
        archivo >> anioEdicion;
        archivo.ignore();
        getline(archivo, editorial, ',');
        getline(archivo, ISBN, ',');
        archivo >> numPaginas;
        archivo.ignore();

        if (titulo != libroSolicitado)
        {
            temp << titulo << "," << autor << "," << anioEdicion << "," << editorial << "," << ISBN << "," << numPaginas << endl;
        }
    }

    archivo.close();
    temp.close();
    remove("biblioteca.txt");
    rename("temp.txt", "biblioteca.txt");
}

bool verificarDNI(const string &dni)
{
    Lector *actual = frente;
    while (actual)
    {
        if (actual->dni == dni)
        {
            return true;
        }
        actual = actual->siguiente;
    }
    return false;
}

void solicitarLibro(const string &dni, const string &nombre, const string &libro)
{
    if (nombre.empty() || dni.empty())
    {
        cout << "Error: Se necesita el nombre y el DNI para continuar." << endl;
        return;
    }

    Libro *actual = cabeza;
    while (actual)
    {
        if (actual->titulo == libro)
        {
            eliminarLibroDeLista(libro);
            eliminarLibroDelArchivo(libro);
            Lector *nuevoLector = new Lector{nombre, dni, libro, nullptr};
            if (!frente)
            {
                frente = final = nuevoLector;
            }
            else
            {
                final->siguiente = nuevoLector;
                final = nuevoLector;
            }
            cout << nombre << " ha solicitado el libro: " << libro << endl;
            agregarHistorial("Solicitar libro: " + libro + " por " + nombre);
            return;
        }
        actual = actual->siguiente;
    }
    cout << "El libro " << libro << " no esta disponible." << endl;
}

void devolverLibro(const string &libro)
{
    Lector *actual = frente;
    Lector *anterior = nullptr;

    while (actual)
    {
        if (actual->libroSolicitado == libro)
        {

            if (anterior)
            {
                anterior->siguiente = actual->siguiente;
            }
            else
            {
                frente = actual->siguiente;
            }
            if (actual == final)
            {
                final = anterior;
            }
            delete actual;

            Libro *nuevoLibro = new Libro{libro, "", 0, "", "", 0, cabeza};
            cabeza = nuevoLibro;
            agregarHistorial("Devolver libro: " + libro);

            ofstream archivo("biblioteca.txt", ios::app);
            archivo << libro << ",,,0,,0\n";
            archivo.close();

            cout << "Libro devuelto: " << libro << endl;
            return;
        }
        anterior = actual;
        actual = actual->siguiente;
    }
    cout << "El libro no esta en solicitudes." << endl;
}

void mostrarSolicitudes()
{
    Lector *actual = frente;
    if (!actual)
    {
        cout << "No hay solicitudes de libros." << endl;
    }
    else
    {
        while (actual)
        {
            cout << "Nombre: " << actual->nombre << ", DNI: " << actual->dni
                 << ", Libro solicitado: " << actual->libroSolicitado << endl;
            actual = actual->siguiente;
        }
    }
}

void buscarLibro(const string &titulo)
{
    Libro *actual = cabeza;
    while (actual)
    {
        if (actual->titulo == titulo)
        {
            cout << "Libro encontrado:\nTitulo: " << actual->titulo
                 << ", Autor: " << actual->autor << ", Anio: " << actual->anioEdicion
                 << ", Editorial: " << actual->editorial << ", ISBN: " << actual->ISBN
                 << ", Paginas: " << actual->numPaginas << endl;
            return;
        }
        actual = actual->siguiente;
    }
    cout << "Libro no encontrado." << endl;
}

void mostrarHistorial()
{
    if (indiceHistorial == 0)
    {
        cout << "No hay acciones en el historial." << endl;
    }
    else
    {
        cout << "Historial de acciones:" << endl;
        for (int i = 0; i < indiceHistorial; ++i)
        {
            cout << historial[i] << endl;
        }
    }
}

void guardarSolicitudes()
{
    ofstream archivoSolicitudes("solicitudes.txt");
    Lector *actual = frente;
    while (actual)
    {
        archivoSolicitudes << "Nombre: " << actual->nombre << ", DNI: " << actual->dni
                           << ", Libro solicitado: " << actual->libroSolicitado << endl;
        actual = actual->siguiente;
    }
    archivoSolicitudes.close();
}

void guardarDatos(const string &nombreArchivoLibros)
{
    ofstream archivoLibros(nombreArchivoLibros);
    Libro *actualLibro = cabeza;
    while (actualLibro)
    {
        archivoLibros << actualLibro->titulo << "," << actualLibro->autor << ","
                      << actualLibro->anioEdicion << "," << actualLibro->editorial << ","
                      << actualLibro->ISBN << "," << actualLibro->numPaginas << endl;
        actualLibro = actualLibro->siguiente;
    }
    archivoLibros.close();
    guardarSolicitudes();
    agregarHistorial("Guardar datos");
    cout << "Datos guardados en archivos." << endl;
}

int main()
{
    cargarLibrosDesdeArchivo("biblioteca.txt");
    int opcion;
    do
    {
        cout << "\n--- Menu ---" << endl;
        cout << "1. Agregar libro" << endl;
        cout << "2. Mostrar libros" << endl;
        cout << "3. Ordenar libros por titulo" << endl;
        cout << "4. Solicitar libro" << endl;
        cout << "5. Devolver libro" << endl;
        cout << "6. Mostrar solicitudes" << endl;
        cout << "7. Buscar libro" << endl;
        cout << "8. Mostrar historial" << endl;
        cout << "9. Guardar datos" << endl;
        cout << "0. Salir" << endl;
        cout << "Seleccione una opcion: ";
        cin >> opcion;
        cin.ignore();

        switch (opcion)
        {
        case 1:
        {
            string titulo, autor, editorial, ISBN;
            int anioEdicion, numPaginas;
            cout << "Ingrese titulo: ";
            getline(cin, titulo);
            cout << "Ingrese autor: ";
            getline(cin, autor);
            cout << "Ingrese anio de edicion: ";
            cin >> anioEdicion;
            cin.ignore();
            cout << "Ingrese editorial: ";
            getline(cin, editorial);
            cout << "Ingrese ISBN: ";
            getline(cin, ISBN);
            cout << "Ingrese numero de paginas: ";
            cin >> numPaginas;
            cin.ignore();
            agregarLibro(titulo, autor, anioEdicion, editorial, ISBN, numPaginas);
            break;
        }
        case 2:
            mostrarLibros();
            break;
        case 3:
            ordenarLibros();
            break;
        case 4:
        {
            string dni, nombre, libro;
            cout << "Ingrese DNI del solicitante: ";
            getline(cin, dni);
            cout << "Ingrese nombre del solicitante: ";
            getline(cin, nombre);
            cout << "Ingrese libro solicitado: ";
            getline(cin, libro);
            solicitarLibro(dni, nombre, libro);
            break;
        }
        case 5:
        {
            string libro;
            cout << "Ingrese el titulo del libro a devolver: ";
            getline(cin, libro);
            devolverLibro(libro);
            break;
        }
        case 6:
            mostrarSolicitudes();
            break;
        case 7:
        {
            string titulo;
            cout << "Ingrese el titulo del libro a buscar: ";
            getline(cin, titulo);
            buscarLibro(titulo);
            break;
        }
        case 8:
            mostrarHistorial();
            break;

        case 9:
            guardarDatos("biblioteca.txt");
            break;
        case 0:
            cout << "Saliendo..." << endl;
            break;
        default:
            cout << "Opcion no valida." << endl;
            break;
        }
    } while (opcion != 0);
    return 0;
}

File: Trabajo Parcial/parcial.py
class Libro:
    def __init__(self, titulo, autor, anio_edicion, editorial, ISBN, num_paginas, reservado=False):
        self.titulo = titulo
        self.autor = autor
        self.anio_edicion = anio_edicion
        self.editorial = editorial
        self.ISBN = ISBN
        self.num_paginas = num_paginas
        self.reservado = reservado
        self.siguiente = None


class Lector:
    def __init__(self, nombre, dni, libro_solicitado):
        self.nombre = nombre
        self.dni = dni
        self.libro_solicitado = libro_solicitado
        self.siguiente = None


# Variables globales
cabeza = None  # Lista de libros
frente = None  # Inicio de la cola de lectores
final = None   # Final de la cola de lectores
historial = []
MAX_HISTORIAL = 100


def agregar_historial(accion):
    if len(historial) < MAX_HISTORIAL:
        historial.append(accion)


def cargar_libros_desde_archivo(nombre_archivo):
    global cabeza
    try:
        with open(nombre_archivo, 'r') as archivo:
            for linea in archivo:
                partes = linea.strip().split(',')
                if len(partes) == 6:
                    titulo, autor, anio_edicion, editorial, ISBN, num_paginas = partes
                    nuevo_libro = Libro(titulo, autor, int(anio_edicion), editorial, ISBN, int(num_paginas))
                    nuevo_libro.siguiente = cabeza
                    cabeza = nuevo_libro
    except IOError:
        print(f"Error: No se pudo abrir el archivo {nombre_archivo}")


def cargar_solicitudes_desde_archivo(nombre_archivo):
    global frente, final
    try:
        with open(nombre_archivo, 'r') as archivo:
            for linea in archivo:
                partes = linea.strip().split(',')
                if len(partes) == 3:
                    nombre, dni, libro = partes
                    nuevo_lector = Lector(nombre, dni, libro)
                    if frente is None:
                        frente = final = nuevo_lector
                    else:
                        final.siguiente = nuevo_lector
                        final = nuevo_lector
    except IOError:
        print(f"Error: No se pudo abrir el archivo {nombre_archivo}")


def agregar_libro(titulo, autor, anio_edicion, editorial, ISBN, num_paginas):
    global cabeza
    nuevo_libro = Libro(titulo, autor, anio_edicion, editorial, ISBN, num_paginas)
    nuevo_libro.siguiente = cabeza
    cabeza = nuevo_libro
    agregar_historial(f"Agregar libro: {titulo}")


def mostrar_libros():
    if cabeza is None:
        print("No hay libros en la biblioteca.")
        return
    actual = cabeza
    while actual:
        estado_reservado = " (Reservado)" if actual.reservado else ""
        print(f"Titulo: {actual.titulo}, Autor: {actual.autor}, Anio: {actual.anio_edicion}, "
              f"Editorial: {actual.editorial}, ISBN: {actual.ISBN}, Paginas: {actual.num_paginas}{estado_reservado}")
        actual = actual.siguiente


def eliminar_libro_del_archivo(libro_solicitado):
    lineas = []
    with open("biblioteca.txt", 'r') as archivo:
        lineas = archivo.readlines()

    with open("temp.txt", 'w') as temp:
        for linea in lineas:
            if libro_solicitado not in linea:
                temp.write(linea)

    import os
    os.remove("biblioteca.txt")
    os.rename("temp.txt", "biblioteca.txt")


def devolver_libro_al_archivo(libro_devuelto):
    with open("biblioteca.txt", 'a') as archivo:
        archivo.write(f"{libro_devuelto.titulo},{libro_devuelto.autor},{libro_devuelto.anio_edicion},"
                       f"{libro_devuelto.editorial},{libro_devuelto.ISBN},{libro_devuelto.num_paginas}\n")


def solicitar_libro(dni, nombre, libro):
    if not nombre or not dni:
        print("Error: Se necesita el nombre y el DNI para continuar.")
        return

    global frente, final
    actual = cabeza
    while actual:
        if actual.titulo == libro:
            if actual.reservado:
                print("Este libro ya fue reservado.")
                return
            actual.reservado = True  # Marcar como reservado
            eliminar_libro_del_archivo(libro)  # Eliminar libro de biblioteca.txt
            nuevo_lector = Lector(nombre, dni, libro)
            if frente is None:
                frente = final = nuevo_lector
            else:
                final.siguiente = nuevo_lector
                final = nuevo_lector
            print(f"{nombre} ha solicitado el libro: {libro}")
            agregar_historial(f"Solicitar libro: {libro} por {nombre}")
            return
        actual = actual.siguiente
    print(f"El libro {libro} no está disponible.")


def devolver_libro():
    global frente
    if frente is None:
        print("No hay solicitudes pendientes.")
        return
    print(f"{frente.nombre} ha devuelto el libro: {frente.libro_solicitado}")

    # Marcar el libro como disponible
    actual = cabeza
    while actual:
        if actual.titulo == frente.libro_solicitado:
            actual.reservado = False  # Desmarcar como reservado
            devolver_libro_al_archivo(actual)  # Agregar libro de nuevo a biblioteca.txt
            break
        actual = actual.siguiente

    agregar_historial(f"Devolver libro: {frente.libro_solicitado} por {frente.nombre}")
    temp = frente
    frente = frente.siguiente

    if frente:
        print(f"Atendiendo a: {frente.nombre} para el libro: {frente.libro_solicitado}")

    del temp


def mostrar_solicitudes():
    actual = frente
    if actual is None:
        print("No hay solicitudes pendientes.")
        return
    while actual:
        print(f"Solicitante: {actual.nombre}, DNI: {actual.dni}, Libro solicitado: {actual.libro_solicitado}")
        actual = actual.siguiente


def mostrar_historial():
    print("Historial de acciones:")
    for accion in historial:
        print(f"- {accion}")


def guardar_datos(nombre_archivo_libros, nombre_archivo_solicitudes):
    with open(nombre_archivo_libros, 'w') as archivo_libros:
        actual_libro = cabeza
        while actual_libro:
            archivo_libros.write(f"{actual_libro.titulo},{actual_libro.autor},{actual_libro.anio_edicion},"
                                 f"{actual_libro.editorial},{actual_libro.ISBN},{actual_libro.num_paginas}\n")
            actual_libro = actual_libro.siguiente

    with open(nombre_archivo_solicitudes, 'w') as archivo_solicitudes:
        actual_lector = frente
        while actual_lector:
            archivo_solicitudes.write(f"{actual_lector.nombre},{actual_lector.dni},{actual_lector.libro_solicitado}\n")
            actual_lector = actual_lector.siguiente

    agregar_historial("Guardar datos")
    print("Datos guardados en archivos.")


def quicksort(libros, izquierda, derecha):
    i = izquierda
    j = derecha
    pivote = libros[(izquierda + derecha) // 2].titulo

    while i <= j:
        while libros[i].titulo < pivote:
            i += 1
        while libros[j].titulo > pivote:
            j -= 1
        if i <= j:
            libros[i], libros[j] = libros[j], libros[i]
            i += 1
            j -= 1

    if izquierda < j:
        quicksort(libros, izquierda, j)
    if i < derecha:
        quicksort(libros, i, derecha)


def ordenar_libros():
    global cabeza
    libros = []
    actual = cabeza
    while actual:
        libros.append(actual)
        actual = actual.siguiente

    quicksort(libros, 0, len(libros) - 1)

    print("Libros ordenados por título:")
    for libro in libros:
        print(libro.titulo)


def busqueda_binaria(libros, titulo_buscado):
    izquierda = 0
    derecha = len(libros) - 1

    while izquierda <= derecha:
        medio = izquierda + (derecha - izquierda) // 2
        if libros[medio].titulo == titulo_buscado:
            return medio  # Título encontrado
        if libros[medio].titulo < titulo_buscado:
            izquierda = medio + 1
        else:
            derecha = medio - 1
    return -1  # Título no encontrado


def buscar_libro(titulo_buscado):
    global cabeza
    libros = []
    actual = cabeza
    while actual:
        libros.append(actual)
        actual = actual.siguiente

    resultado = busqueda_binaria(libros, titulo_buscado)
    if resultado != -1:
        print(f"Libro encontrado: {libros[resultado].titulo}")
    else:
        print(f"El libro {titulo_buscado} no se encuentra en la biblioteca.")


def main():
    cargar_libros_desde_archivo("biblioteca.txt")
    cargar_solicitudes_desde_archivo("solicitudes.txt")

    while True:
        print("\nMenú:")
        print("1. Agregar libro")
        print("2. Mostrar libros")
        print("3. Solicitar libro")
        print("4. Devolver libro")
        print("5. Mostrar solicitudes")
        print("6. Mostrar historial")
        print("7. Guardar datos")
        print("8. Ordenar libros")
        print("9. Buscar libro")
        print("0. Salir")

        opcion = input("Seleccione una opción: ")
        if opcion == '1':
            titulo = input("Título: ")
            autor = input("Autor: ")
            anio_edicion = int(input("Año de edición: "))
            editorial = input("Editorial: ")
            ISBN = input("ISBN: ")
            num_paginas = int(input("Número de páginas: "))
            agregar_libro(titulo, autor, anio_edicion, editorial, ISBN, num_paginas)
        elif opcion == '2':
            mostrar_libros()
        elif opcion == '3':
            nombre = input("Nombre del solicitante: ")
            dni = input("DNI del solicitante: ")
            libro = input("Título del libro solicitado: ")
            solicitar_libro(dni, nombre, libro)
        elif opcion == '4':
            devolver_libro()
        elif opcion == '5':
            mostrar_solicitudes()
        elif opcion == '6':
            mostrar_historial()
        elif opcion == '7':
            guardar_datos("biblioteca.txt", "solicitudes.txt")
        elif opcion == '8':
            ordenar_libros()
        elif opcion == '9':
            titulo_buscado = input("Título del libro a buscar: ")
            buscar_libro(titulo_buscado)
        elif opcion == '0':
            break
        else:
            print("Opción no válida.")


if __name__ == "__main__":
    main()

